client/extensions/woocommerce/app/settings/shipping/shipping-class-dialog.js

```diff
@@ -0,0 +1,218 @@
+				onClick: this.onSave,
+				isPrimary: true,
+				disabled: ! canSave,
+			},
+		];
+
+		if ( ! isNew ) {
+			buttons.unshift( {
+				action: 'delete',
+				label: (
+					<span>
+						<Gridicon icon="trash" /> { translate( 'Delete this class' ) }
+					</span>
+				),
+				onClick: this.onDelete,
+				additionalClassNames: 'shipping__class-delete is-scary is-borderless',
+			} );
+		}
+
+		const dialogProps = {
+			additionalClassNames: 'shipping__class-dialog woocommerce',
+			isVisible: true,
+			buttons: buttons,
+			onClose: this.onCancel,
+		};
+
+		return (
+			<Dialog { ...dialogProps }>
+				<div className="shipping__class-dialog-header">
+					<h3 className="shipping__class-dialog-heading">
+						{ isNew ? translate( 'Add shipping class' ) : translate( 'Edit shipping class' ) }
+					</h3>
+				</div>
+
+				<FormFieldSet className="shipping__class-dialog-name">
+					<FormLabel>{ translate( 'Name' ) }</FormLabel>
+					<FormTextInput type="text" name="name" value={ data.name } onChange={ this.onChange } />
+				</FormFieldSet>
+
+				<FormFieldSet className="shipping__class-dialog-slug">
+					<FormLabel>{ translate( 'Slug' ) + ': ' }</FormLabel>
+					<FormClickToEditInput
+						onChange={ this.onSlugChange }
+						value={ this.getSlugToDisplay() }
+						placeholder="-"
+						editAriaLabel={ translate( 'Edit slug' ) }
+						updateAriaLabel={ translate( 'Update slug' ) }
+					/>
+				</FormFieldSet>
+
+				<FormFieldSet>
+					<FormLabel>{ translate( 'Description' ) }</FormLabel>
+					<FormTextInput
+						type="text"
+						name="description"
+						value={ data.description }
+						onChange={ this.onChange }
+					/>
+				</FormFieldSet>
+			</Dialog>
+		);
+	}
+
+	getSlugToDisplay() {
+		const { name, slug } = this.props.data;
+
+		return slug.length ? slug : name.toLowerCase().replace( /\s+/g, '-' );
+	}
+
+	onChange = ( { target: { name, value } } ) => {
+		const { siteId, updateFieldValue } = this.props;
+
+		updateFieldValue( siteId, name, value );
+	};
+
+	onSlugChange = value => {
+		const { siteId, updateFieldValue } = this.props;
+
+		updateFieldValue( siteId, 'slug', value );
+	};
+
+	onCancel = () => {
+		const { close, siteId } = this.props;
+
+		close( siteId );
+	};
+
+	onDelete = () => {
+		const {
+			data: { id },
+			siteId,
+			deleteClass,
+		} = this.props;
+
+		deleteClass( siteId, id );
+	};
+
+	onSave = () => {
+		const {
+			occupiedNames,
+			occupiedSlugs,
+			siteId,
+			translate,
+			data: { name, slug },
+			updateFieldValue,
+			save,
+		} = this.props;
+
+		const slugToUse = this.getSlugToDisplay();
+
+		// Check if the name is free
+		if ( -1 !== occupiedNames.indexOf( name ) ) {
+			return this.props.errorNotice(
+				translate( 'The name "%(name)s" is already occupied.', {
+					args: { name },
+				} )
+			);
+		}
+
+		if ( -1 !== occupiedSlugs.indexOf( slugToUse ) ) {
+			return this.props.errorNotice(
+				translate( 'The slug "%(slug)s" is already occupied.', {
```

Comment:
Is there another phrasing we can use besides "occupied", like "used" or "existing"? At least in my experience, it's not a common term to refer to such key collisions. (From the Edit Product screen: https://cloudup.com/c6XP5Z7Aq4d)

Also, the notice can accept an ID, which I think is appropriate here as one notice would then obsolete the last (and can also be cleared upon successful save.)

Lastly, the argument here needs to be adjusted to the variable name change from just now 🙂

Reply:
Done, done and done.